File: python-src/detect_road.py
import cv2
import numpy as np
import math

cap = cv2.VideoCapture('sample_vid/road1.avi')
fourcc = cv2.VideoWriter_fourcc(*'XVID')
fps = 30
width = int(cap.get(3))
height = int(cap.get(4))
frame_size = (width, height)
out = cv2.VideoWriter('sample_vid/output.avi',fourcc, fps, frame_size)


# Region of Interest
roiWidth = width
roiHeight = height
roiX = 0
roiY = int(height * 2/3)

#Process for lines on the road
def slope(vx1, vx2, vy1, vy2):         #Parameters to calculate slope
    m=float(vy2-vy1)/float(vx2-vx1)        #Slope equation
    theta = math.atan(m)                  #calculate the slope angle
    return theta*(180/np.pi)              #Calculated angle in radians

# Image processing
while (cap.isOpened()):
    ret, frame = cap.read()
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)            

    #Using histogram equalization to increase image contrast
    hist = cv2.equalizeHist(gray)                              
    video = cv2.cvtColor(gray,cv2.COLOR_GRAY2BGR)
    cv2.rectangle(video, (roiX, roiY), (width, height), (255, 0, 0),1) 

    # Change color space to extract lane color
    hsv = cv2.cvtColor(video, cv2.COLOR_BGR2HSV)              
    grayROI  = hsv[roiY:roiY+roiHeight, roiX:roiX+roiWidth]  
    low_white = np.array([0,0,90],dtype=np.uint8)            
    high_white = np.array([255,255,255],dtype=np.uint8)      
    mask = cv2.inRange(grayROI,low_white,high_white)          


    #Length for lines to find
    minLineLength = 100
    maxLineGap = 40
    lines = cv2.HoughLinesP(mask,1,np.pi/180,10,minLineLength,maxLineGap)
    frameClone = frame.copy()
    for x1,y1,x2,y2 in lines[0]:
      if (round(x2-x1)!=0):
            arctan = slope(x1,x2,y1,y2)
            if (round(arctan>=round(-80)) and round(arctan<=round(-30))):

                cv2.rectangle(frameClone, (x1 + roiX, y1 + roiY), (x2 + roiX, y2 + roiY), (0, 255, 0), 3, lineType=8, shift=0)
            if (round(arctan>=round(30)) and round(arctan<=round(80))):
                #cv2.line(frameClone,(x1 + roiX,y1 + roiY),(x2 + roiX,y2 + roiY),(0, 255, 0),2)
                cv2.rectangle(frameClone, (x1 + roiX, y1 + roiY), (x2 + roiX, y2 + roiY), (0, 255, 0), 3, lineType=8, shift=0)
                #print x1 + roiX

    cv2.imshow('Video', frameClone)
    out.write(frameClone)
    if cv2.waitKey(30) & 0xFF == ord('q'):
        break


cap.release()
out.release()
cv2.destroyAllWindows()

File: detect_road.cpp
#include <iostream>
#include <math.h>
#include "opencv_tools.h"

using namespace cv;
using namespace std;

int main()
{        
    VideoCapture cap("sample_vid/road1.avi");
    if (!cap.isOpened()) {
        return -1;
    }

    int frame_width = cap.get(CAP_PROP_FRAME_WIDTH);
    int frame_height = cap.get(CAP_PROP_FRAME_HEIGHT);

    // Region of Interest
    cv::Rect roi;
    roi.x = 0;
    roi.y = int(frame_height * 3/4);
    roi.width = frame_width;
    roi.height = frame_height - roi.y;
    
    for(;;)
    {
        Mat frame, mask, denoised;
        Mat gray, hist, video, hsv, grayROI;
        cap >> frame; // get a new frame from camera
        cvtColor(frame, gray, COLOR_BGR2GRAY);
        GaussianBlur(gray, gray, Size(7,7), 1.5, 1.5);

        // increase contrast
        equalizeHist(gray, hist);
        //add_gaussian_noise(gray);
        //medianBlur(gray, gray, 5);


        // add boundary for region of interest 
        cvtColor(gray, video, COLOR_GRAY2BGR);
        rectangle(video, Point(roi.x, roi.y), Point(roi.width, roi.y + roi.height), Scalar(0,255,0), 1);

        // Change color space to extract lane color
        cvtColor(video, hsv, COLOR_BGR2HSV);
        grayROI = hsv(roi);
        inRange(grayROI, Scalar(0,0,90), Scalar(255, 255,255), mask);

        // Length for lines to find
        int minLineLength = 30;
        int maxLineGap = 10;
        cv::Mat frameClone(frame);
        vector<Vec4i> lines;
        HoughLinesP(mask, lines, 1, CV_PI/180, 30, minLineLength, maxLineGap);

        Vec4i lineL = Vec4i(0, 0, 0, 0);
        Vec4i lineR = Vec4i(0, 0, 0, 0);

        for (int i = 0; i < lines.size(); ++i) {
            Vec4i l = lines[i];
            
            float x1 = l[0];
            float y1 = l[1];
            float x2 = l[2];
            float y2 = l[3];
            if (round(x2-x1) != 0) {
                float arctan = slope(x1, x2, y1, y2);
                float realX1 = x1 + roi.x;
                float realX2 = x2 + roi.x;
                float realY1 = y1 + roi.y;
                float realY2 = y2 + roi.y;
                if (realX1 < frame_width / 2 && realX2 < frame_width / 2)
                //if (round(arctan>=round(-150)) && round(arctan<=round(50))) 
                    if (distance(realX1, realX2, realY1, realY2) 
                            > get_length(lineL)) {
                        lineL = Vec4i(realX1, realX2, realY1, realY2);
                    }
                if (realX1 >= frame_width / 2 && realX2 >= frame_width / 2)
                //if (round(arctan>=round(20)) &&  round(arctan<=round(150)))
                    if (distance(realX1, realX2, realY1, realY2) 
                            > get_length(lineR)) {
                        lineR = Vec4i(realX1, realX2, realY1, realY2);
                    }
            }
        }

        // drawing on original frame
        draw_line(frameClone, lineL);
        draw_line(frameClone, lineR);
        //draw_rect(frameClone, lineL);
        //draw_rect(frameClone, lineR);
        Point midpointL = get_midpoint(lineL);
        Point midpointR = get_midpoint(lineR);
        Point centerP = get_midpoint(midpointL, midpointR);
        draw_line(frameClone, midpointL, midpointR);
        draw_circle(frameClone, centerP);

        std::cout << centerP.x << " " << centerP.y << std::endl;

        imshow("frame", frameClone);
        waitKey(0);
    }
    
    cap.release();
    return 0;
}
